src/core/agent/services/ipexCommunicationService.ts

```diff
@@ -0,0 +1,266 @@
+    this.eventService.on(
+      AcdcKeriEventTypes.AcdcKeriStateChanged,
+      async (event: AcdcKeriStateChangedEvent) => {
+        callback(event);
+      }
+    );
+  }
+
+  async acceptKeriAcdc(id: string): Promise<void> {
+    const keriNoti = await this.getKeriNotificationRecordById(id);
+    const keriExchange = await this.signifyClient
+      .exchanges()
+      .get(keriNoti.a.d as string);
+    const credentialId = keriExchange.exn.e.acdc.d;
+    await this.createAcdcMetadataRecord(keriExchange.exn);
+
+    this.eventService.emit<AcdcKeriStateChangedEvent>({
+      type: AcdcKeriEventTypes.AcdcKeriStateChanged,
+      payload: {
+        credentialId,
+        status: CredentialStatus.PENDING,
+      },
+    });
+    let holderSignifyName;
+    const holder = await this.identifierStorage.getIdentifierMetadata(
+      keriExchange.exn.a.i
+    );
+    if (holder && holder.signifyName) {
+      holderSignifyName = holder.signifyName;
+    } else {
+      const identifierHolder = await this.getIdentifierById(
+        keriExchange.exn.a.i
+      );
+      holderSignifyName = identifierHolder?.name;
+    }
+    if (!holderSignifyName) {
+      throw new Error(IpexCommunicationService.ISSUEE_NOT_FOUND);
+    }
+
+    await this.admitIpex(
+      keriNoti.a.d as string,
+      holderSignifyName,
+      keriExchange.exn.i
+    );
+
+    // @TODO - foconnor: This should be event driven, need to fix the notification in KERIA/Signify.
+    const cred = await this.waitForAcdcToAppear(credentialId);
+    const credentialShortDetails = await this.updateAcdcMetadataRecordCompleted(
+      credentialId,
+      cred
+    );
+    await this.deleteKeriNotificationRecordById(id);
+    this.eventService.emit<AcdcKeriStateChangedEvent>({
+      type: AcdcKeriEventTypes.AcdcKeriStateChanged,
+      payload: {
+        status: CredentialStatus.CONFIRMED,
+        credential: credentialShortDetails,
+      },
+    });
+  }
+
+  private async waitForAcdcToAppear(credentialId: string): Promise<any> {
+    let acdc = await this.getCredentialBySaid(credentialId);
+    let retryTimes = 0;
+    while (!acdc) {
+      if (retryTimes > 120) {
+        throw new Error(IpexCommunicationService.ACDC_NOT_APPEARING);
+      }
+      await new Promise((resolve) => setTimeout(resolve, 500));
+      acdc = (await this.getCredentialBySaid(credentialId)).acdc;
+      retryTimes++;
+    }
+    return acdc;
+  }
+
+  private async getKeriNotificationRecordById(
+    id: string
+  ): Promise<KeriNotification> {
+    const result = await this.basicStorage.findById(id);
+    if (!result) {
+      throw new Error(
+        `${IpexCommunicationService.KERI_NOTIFICATION_NOT_FOUND} ${id}`
+      );
+    }
+    return {
+      id: result.id,
+      createdAt: result.createdAt,
+      a: result.content,
+    };
+  }
+
+  async deleteKeriNotificationRecordById(id: string): Promise<void> {
+    await this.basicStorage.deleteById(id);
```

Comment:
This can be moved into `acceptKeriAcdc` - it's too small to be a tested function